### PR TITLE
Cap JSON request bodies before parsing (readJsonWithLimit across all JSON routes)

### LINKED ISSUE #264: JSON request bodies are parsed before any size limit applies
## The gap

Every JSON route calls `c.req.json()` with no ceiling on the bytes it will parse. There is no body-limit middleware in `src/middleware/`, and no route sets one:

```
$ grep -rn "bodyLimit\|Content-Length\|MAX_BODY" src/middleware/ src/index.ts
(no matches)
```

The workspace commit route (`src/routes/workspaces.ts`) is the clearest case. It enforces a 2000-file count, a 10 MiB per-file cap, and a 25 MiB aggregate — but every one of those runs **after** the parse:

```ts
const body = await c.req.json<{ files?: unknown; message?: unknown; projectId?: unknown }>();
// … only now do the size checks run
```

So a 400 MB JSON body is fully materialized in the isolate before the code that exists to bound memory gets a chance to reject it. The caps protect the clone and the commit; they do not protect the parse.

The git-receive-pack path already does this correctly — it reads the body with a 50 MiB ceiling and inspects a bounded 1 MiB view — so the pattern exists in the codebase, just not on the JSON routes.

## Why it is filed rather than fixed

Surfaced during review of #216, whose scope is the S3/S5/S6/S7 items of the workspace-write hardening analysis. That PR tightened the post-parse budget (per-file cap enforced before cloning, path keys and message now counted). Capping the pre-parse body is a different defense at a different layer: it needs a shared read-with-limit helper applied across the JSON routes, not a change to one validation loop, and it should be decided once for the whole surface rather than bolted onto the route that happened to be under review.

## Suggested shape

- A small `readJsonWithLimit(c, maxBytes)` helper: read the raw body with a byte ceiling, reject with 413 past it, then parse.
- Apply to the workspace commit route first (largest documented payload, 25 MiB), then the other write routes.
- Pick the ceiling relative to the isolate's ~128 MB budget, leaving room for the parsed object graph — the parsed form of a JSON string is larger than its wire bytes.

## Prior art in-repo

`src/routes/git-http.ts` — `MAX_COMMAND_SECTION_BYTES` and the 50 MiB body ceiling, including the bounded-inflate path that avoids trusting a compressed length.

## Notes

Pre-release, no users, so this is hardening rather than an incident. Raised by CodeRabbit on #216 as part of a broader finding; the post-parse half of that finding is fixed in `dfcdc63` and `6bd16ac`.

### PR DESCRIPTION
**Pull request by jlamoreaux**
## Summary

Every JSON route parsed the full request body before any size limit applied — the workspace-commit route's post-parse caps (2000 files, 10 MiB/file, 25 MiB aggregate) only ran after a potentially 400 MB body was already fully materialized in the isolate.

This adds a shared `readJsonWithLimit(c, maxBytes)` helper (`src/utils/request-body.ts`) that streams the body and aborts once bytes read exceed the cap, mirroring the bounded-read pattern the git-receive-pack path already uses in `src/routes/git-http.ts`. A declared `Content-Length` over the cap is a cheap early reject, but enforcement never trusts it — absent or understated lengths are caught mid-stream.

- Workspace commit route: 32 MiB ceiling (`MAX_COMMIT_BODY_BYTES`, above its 25 MiB post-parse aggregate).
- Every other `c.req.json()` call site (22 across 10 route files): 1 MiB default, each as its own named constant.
- New `PayloadTooLargeError` (413, `PAYLOAD_TOO_LARGE`) in the existing `AppError` taxonomy, with a `payloadTooLarge()` response helper producing the standard `{ error, code }` shape.
- No post-parse validation logic changed. Malformed JSON within the cap still throws exactly like `c.req.json()`, so existing `.catch(() => ({}))` call sites behave identically.

Note: this sweeps many route files, so it will conflict textually (small, mechanical) with in-flight PRs touching the same routes (e.g. #216).

## Related issues

Closes #264

## Type of change

- [x] Bug fix

## How was this verified?

- [x] `npm run typecheck`
- [x] `npm test` (full suite: 1733 passed, no regressions)
- [x] `npm run lint`

- New `tests/request-body.test.ts` (8 tests): declared Content-Length over cap rejects before the stream is read (asserted via `stream.locked`); no-Content-Length body still capped mid-stream and never parsed; understated Content-Length doesn't bypass the cap; 413 body shape; exact-boundary cases; malformed-JSON-within-cap still throws.
- New `tests/workspaces-commit-body-limit.test.ts`: a 33 MiB streamed body with no Content-Length hits the real 32 MiB route ceiling → 413 and `commitAndPush` never runs; a normal small commit still succeeds.

## Checklist

- [x] Tests added or updated for the change (coverage thresholds still pass)
- [x] Docs updated where the change makes them inaccurate (none affected)
- [x] No secrets, tokens, or personal data committed
- [x] The web UI change (if any) stays server-rendered with no client-side JavaScript (no UI change)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01Gk1wNDsuwpt6q36oZaZhof

---
_Generated by [Claude Code](https://claude.ai/code/session_01Gk1wNDsuwpt6q36oZaZhof)_

## Summary by CodeRabbit

* **Bug Fixes**
  * Added request-body size limits across API operations to prevent oversized requests from being processed.
  * Oversized payloads now receive consistent HTTP 413 responses with a machine-readable error code.
  * Supports payloads up to 32 MiB for workspace commits and conflict resolution while maintaining smaller limits elsewhere.

* **Tests**
  * Added coverage for streaming requests, size boundaries, malformed JSON, and normal parsing.
  * Added integration tests confirming oversized requests are rejected before processing.

### CODE REVIEW

**Comment by jlamoreaux** [comment]
@coderabbitai review

This PR has not been reviewed yet — earlier attempts were rejected while the hourly allowance was exhausted. CI is green on `34cfdc6` and it merges cleanly into `main`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/routes/issues.ts [hunk 1]
```diff
@@ -30,6 +31,10 @@ const app = new Hono<{ Bindings: Env }>();
 
 const MAX_TITLE_LENGTH = 200;
 const MAX_BODY_LENGTH = 20_000;
+// Issue title/body are short text; cap the raw read well above MAX_BODY_LENGTH
+// (which truncates post-parse) but far under the isolate budget.
+const MAX_ISSUE_BODY_BYTES = 1024 * 1024;
+const MAX_ISSUE_UPDATE_BODY_BYTES = 1024 * 1024;
```
[the remark this answers is not in the document]
Comment (jlamoreaux) [reply src/routes/issues.ts]: Fixed in `6abf7b7`, and worth being precise about the severity, because the regression test shows it exactly. Reverting the route to `await c.req.json<typeof body>()`:

```
 × rejects a body over the 1 MiB pre-parse cap with 413 before parsing — even with no Content-Length header
   → expected 400 to be 413
 × rejects an oversized body on the declared-Content-Length fast path with 413
   → expected 201 to be 413
```

That **`201`** is the vulnerability in miniature: without the cap, an oversized body is fully buffered, parsed, and *written to the database* — truncated to `MAX_COMMENT_LENGTH` only afterwards, which is exactly the post-parse gap you identified. The `400` is the same path with non-JSON filler: buffered and parsed, then discarded.

The fix mirrors the open-issue route at line 160 — a new `MAX_COMMENT_BODY_BYTES` (1 MiB), `readJsonWithLimit`, and the `instanceof Response` guard before validation.

Worth recording why this one escaped the sweep: the call was `c.req.json<typeof body>()` **with a type parameter**, so grepping `c.req.json()` returns nothing. Searching `req\.json` instead finds it — and confirms it was the only uncapped site left. A re-grep after the fix returns two hits, both inside prose comments in `request-body.ts`.

`tests/issues-comment-body-limit.test.ts` uses a real SQLite D1, the real auth middleware, and real project/issue lookup, with only `addIssueComment` spied, covering both enforcement paths plus an under-cap control. One thing caught while writing it: `new Request(...)` does **not** populate `Content-Length`, so the "declared" test was silently exercising the streaming path. The header is now set explicitly and asserted, so the two cannot collapse into duplicates — they log `request body exceeds cap` and `request body exceeds cap (Content-Length)` respectively.

Two notes on scope. I also fixed `MAX_PROJECT_DELETE_BODY_BYTES` in `projects.ts`, which carried the byte-identical "what"-style comment to the `users.ts` one you flagged — fixing one and leaving its twin would have been incoherent. I left the remaining `1024 * 1024` comments alone (`orgs.ts`, `agents.ts`, `webhooks.ts`, `bulk-import.ts`): "a handful of short strings" already conveys the rationale, and rewriting every comment in the PR wasn't the finding.

Gate: typecheck clean, lint clean (305 files), full suite 2104 passed / 27 skipped / 0 failures.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/workspaces-commit-body-limit.test.ts [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+import { commitAndPush } from "../src/storage/git-ops";
+
+const ARTIFACTS_REMOTE = "https://acct.artifacts.cloudflare.net/git/@owner/repo.git";
+const WS_REMOTE = "https://acct.artifacts.cloudflare.net/git/@owner/myws.git";
+
+/**
+ * The commit route resolves the project and workspace through KV before it
+ * ever looks at the body, so the test needs that lookup to succeed — otherwise
+ * a 404 would pass for a "rejection" and the cap would go untested.
```
[the remark this answers is not in the document]
Comment (jlamoreaux) [reply tests/workspaces-commit-body-limit.test.ts]: You're right, and the comments were mine — fixed in `a87b431`.

Verified the ordering rather than taking it on faith this time: `readJsonWithLimit` is `workspaces.ts:229`, `getWorkspace` is `283`, `getProjectById` is `305`. The cap runs first, so an oversized body is rejected before the route looks up anything.

That inverts the rationale I had written. I claimed the fixtures were what let the request *reach* the cap; in fact the oversized cases never touch KV at all. The fixtures exist so the **under-cap control case** can get past lookup and show the guard doesn't break an ordinary commit — which is exactly the "keep the KV fixtures for the successful commit control case" you noted, so they stay.

Both docstrings now state the real order, with the line numbers in them. I put the numbers there deliberately: a rationale nobody can check is how the wrong one survived being written in the first place.

Worth owning that this is the same failure mode I've been declining bot findings for all night — asserting a plausible-sounding mechanism without reading the code. Caught here by the review rather than by me.

Gate: typecheck clean, lint clean (305 files), full suite 2104 passed / 27 skipped / 0 failures.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by jlamoreaux** [comment]
@coderabbit review
